File: unity/ProfitableCore/Simulation/PurchaseShip.cs
using Profitable.Core.Constants;
using Profitable.Core.Schema;

namespace Profitable.Core.Simulation;

// Ports src/ships/purchaseShip.ts. Checks wallet sufficiency before
// deducting (a check PurchaseListing never added). No capacity check
// exists here, unlike HireCrew -- the design doc never decided a
// ship-ownership limit analogous to CrewCapacity, so none is enforced.
public static class ShipPurchaser
{
    public static PurchaseShipResult PurchaseShip(ShipCandidate candidate, ShipyardPool pool, Wallet wallet, string playerId)
    {
        if (!pool.AvailableShips.Any(entry => entry.Id == candidate.Id))
        {
            return new PurchaseShipRejected { Reason = "candidate is not in this planet's shipyard pool" };
        }

        if (!ShipsAndTravelConfig.ShipPurchaseCostByTier.TryGetValue(candidate.Tier, out var cost))
        {
            throw new InvalidOperationException($"no purchase cost defined for tier {candidate.Tier}");
        }
        if (wallet.Credits < cost)
        {
            return new PurchaseShipRejected { Reason = $"insufficient funds: need {cost}, have {wallet.Credits}" };
        }

        // A freshly-purchased ship starts with a full tank at its own
        // tier's capacity.
        var fuelCapacity = FuelCapacityDeriver.DeriveFuelCapacity(candidate.Tier);

        var ship = new Ship
        {
            Id = candidate.Id,
            Name = candidate.Name,
            OwnerId = playerId,
            Tier = candidate.Tier,
            CurrentPlanetId = pool.PlanetId,
            FuelCapacity = fuelCapacity,
            CurrentFuel = fuelCapacity,
            Components = candidate.Components,
        };

        var updatedPool = new ShipyardPool
        {
            PlanetId = pool.PlanetId,
            AvailableShips = pool.AvailableShips.Where(entry => entry.Id != candidate.Id).ToList(),
            LastRefreshedAt = pool.LastRefreshedAt,
        };

        return new PurchaseShipSucceeded
        {
            Ship = ship,
            UpdatedPool = updatedPool,
            UpdatedWallet = new Wallet { PlayerId = wallet.PlayerId, Credits = wallet.Credits - cost },
        };
    }
}

File: src/ships/purchaseShip.ts
import type { ShipCandidate } from "../data/types/shipCandidate.ts";
import type { ShipyardPool } from "../data/types/shipyardPool.ts";
import type { Ship } from "../data/types/ship.ts";
import type { Wallet } from "../data/types/wallet.ts";
import type { PurchaseShipResult } from "../data/types/purchaseShipResult.ts";
import { SHIP_PURCHASE_COST_BY_TIER } from "../data/constants/shipsAndTravelConfig.ts";
import { deriveFuelCapacity } from "./deriveFuelCapacity.ts";

// Phase 5 GDD §2.2. Necessary completion: takes the actual candidate/
// pool/wallet data directly rather than IDs into an implicit store --
// same pure-function reasoning already applied to Agent 11's
// purchaseListing() and Agent 16's hireCrew(). Also checks wallet
// sufficiency before deducting (a check Agent 11's purchaseListing()
// never added -- a known, separately-documented gap there); "deducts the
// appropriate cost" only makes sense if there's something to deduct from.
//
// No capacity check exists here, unlike hireCrew() -- the design doc
// never decided a ship-ownership limit analogous to CrewCapacity, so
// none is enforced.
export function purchaseShip(
  candidate: ShipCandidate,
  pool: ShipyardPool,
  wallet: Wallet,
  playerId: string,
): PurchaseShipResult {
  if (!pool.availableShips.some((entry) => entry.id === candidate.id)) {
    return { purchased: false, reason: "candidate is not in this planet's shipyard pool" };
  }

  const cost = SHIP_PURCHASE_COST_BY_TIER.find((entry) => entry.tier === candidate.tier)?.cost;
  if (cost === undefined) {
    throw new RangeError(`no purchase cost defined for tier ${candidate.tier}`);
  }
  if (wallet.credits < cost) {
    return { purchased: false, reason: `insufficient funds: need ${cost}, have ${wallet.credits}` };
  }

  // Ship Fuel amendment: a freshly-purchased ship starts with a full tank
  // at its own tier's capacity -- never STARTING_SHIP_FUEL_CAPACITY, which
  // is exclusively the very first, pre-assigned starting ship's bootstrap
  // value, not a general "new ship" rule.
  const fuelCapacity = deriveFuelCapacity(candidate.tier);

  const ship: Ship = {
    id: candidate.id,
    name: candidate.name,
    ownerId: playerId,
    tier: candidate.tier,
    currentPlanetId: pool.planetId,
    fuelCapacity,
    currentFuel: fuelCapacity,
    components: candidate.components,
  };

  return {
    purchased: true,
    ship,
    updatedPool: {
      ...pool,
      availableShips: pool.availableShips.filter((entry) => entry.id !== candidate.id),
    },
    updatedWallet: { ...wallet, credits: wallet.credits - cost },
  };
}
